### DISCUSSION TITLE
anycable-client remove "on" events

### DISCUSSION THREAD

**Discussion by Renmang** [anycable/anycable#148] [Q&A] [answered]
Hello everyone! Im using anycable-client library in React, and i have this case:

```javascript
useEffect(() => {
	const channel = consumer.subscribeTo('CardChannel', { id });
	channel.on('message', handleMessage);
	return () => {
		channel.disconnect();
	};
}, [consumer, id]);
```
We have 2 active tabs which uses the same consumer, and creates up to 30 channels (one per card, 60 max with both tabs).
Sometimes one card could be in both tabs, so the subscribeTo returns the same channel instance. That works perfect, but when a card is unmounted and the other with same id is still rendered, when the server sends a message, i can see boths channel 'on' events triggered but one fails to do update data as there is no component anymore.
Should the `channel.disconnect()` take care about removing the event?

Another question:
Im using `createCable(url, { channelsCache })`.  When i do the `channel.disconnect()` if that returns true, should i call  `consumer.cache.delete('CardChannel', { id })` myself?

Thank you all in advance

**Comment by palkan** [comment, marked as answer]
> Should the channel.disconnect() take care about removing the event?

Nope. We do not make any assumptions regarding when to remove event listeners; you might want to reconnect the same channel instance later again and keep your listeners. Thus, you need to take care of removing listeners yourself. In your case:

```js
let unbindMessage = channel.on('message', handleMessage);
return () => {
  unbindMessage();
  channel.disconnect();
};
```

> When i do the channel.disconnect() if that returns true, should i call consumer.cache.delete('CardChannel', { id }) myself?

Yes. See this note from the Readme:

> Please, keep in mind, that cache is not cleared automatically. That's your responsibility...

**Reply by Renmang** [reply]
Thank you! About the cache, the delete method uses the identifier key literally:

```javascript
  delete(identifier, params) {
    let key = keyGenerator(identifier, params)

    if (this.store[key]) {
      delete this.store[key]
```

But the cache store has items with `__ghost__` as the identifier. At the begining i was just sending my 'CardChannel' identifier and it wasn't deleting anything... i had to change it to `__ghost__` + the card id.

I didnt dig into why my identifier was changing to "__ghost__" in the cache, but maybe the `delete` method should be aware of that.


Thank you for your attention! You are doing a really great work

**Reply by palkan** [reply]
Hm, I think, that's actually a bug; the channel name should be used a the key. I will take a look (and fix if necessary)

**Reply by palkan** [reply]
Fixed in 0.3.3.

Now `cache.delete("CardChannel", {id})` should work.

**Reply by Renmang** [reply]
Thank you!! i will update
